aiidalab/app.py

```diff
@@ -157,9 +158,15 @@ def installed_version(self):
             return self.metadata.get("version", AppVersion.UNKNOWN)
         return AppVersion.NOT_INSTALLED
 
-    def available_versions(self):
+    def available_versions(self, python_bin=None):
+        """Return a list of available versions excluding the ones with core dependency conflicts."""
         if self.is_registered:
-            yield from sorted(self.releases, key=parse, reverse=True)
+            for version in sorted(self.releases, key=parse, reverse=True):
+                version_requirements = self.releases[version].get("environment")[
```

Comment:
You're using get here but not checking for None. Is it safe to assume that 'environment' and 'python_requirements` are always in the app registry for a given release? Is it robust if an app doesn't define any python requirements?

Reply:
I think if the python requirements are not defined the `version_requirements` will be `None` and it is okay to pass the `None` to function `_strict_dependencies_met`. That simply means there are not dependencies required and the strict dependencies will meet anyway.

Reply:
@unkcpz  I still think this code is backwards. Here's what will happen with the current code if

 - If the `environment` key is not there, the get() call will produce `None["python_requirements"]` which will raise.
 - If `python_requirements` are missing, we raise again since we're not using get()
 
I think the code should instead be (if we're assuming that the `environment` key is always there).
 ```suggestion
                version_requirements = self.releases[version]["environment"].get("python_requirements")
```

Please actually test this by modifying the mocked registry file in the test suite (perhaps add another entry with empty python requirements). Looking at the `strict_dependencies_met` function, I think if you pass it None it will raise as well since it is iterating over the requirements without first checking if they are None.

Reply:
I tried but I don't think it is a good idea to create a "fake" mocked registry entity. I don't know if the aiidalab registry already handle the case if there is nothing in the requirement and in that case the `environment` will be an empty dict or will just not appeared in the registry file. 
Instead of adding a mock test, I give the default value for every attribute get. by `.get("environment", {}).get("python_requirments", [])`. I think this will make things more safe even we don't have unittest for that.

Reply:
Fair enough, but I think we should be proactive here and actually find out what happens when an app doesn't declare any dependencies, because that's something that can very well happen. Will it crash the registry or will it be fine? What would the resulting json file look like? Once we know that, I would add that test case here, there's nothing fake about it.

Reply:
I'll try the mock test, since I find the app registry generator code also here. I think we can test them together from a mocked setup.cfg content.

Reply:
I have a close look at the registry tool, it is safe to get both `environment` and `python_requirement`.
For `environment` it is the key defined in https://github.com/aiidalab/aiidalab/blob/3e8b43135f605cb2af3af25bbbd131da815d0ab9/aiidalab/utils.py#L101-L104.
For `python_requirement`, it is the key mandatory exist in https://github.com/aiidalab/aiidalab/blob/3e8b43135f605cb2af3af25bbbd131da815d0ab9/aiidalab/environment.py#L59-L68

Reply:
I will not work on filling the test gap at the moment. I'll open an issue for doing it in the future.